### ISSUE TITLE
SDL3: Input freezes when rumbling consistently.

### ISSUE THREAD

**Issue by isXander** [libsdl-org/SDL#9441]
## Scenario

Every in-game tick of 20tps I update the controller input using `SDL_GetGamepadAxis` and similar.

My game implements rumble. On the same game tick loop, the game checks if the RumbleManager (responsible for creating patterns etc) has a new rumble state for this tick, if it does, it calls
`SDL_RumbleGamepad(ptrGamepad, state.strong(), state.weak(), 0)`. If it doesn't, it calls the same, but with a state of [0,0].

## Observation

In this game, the rumble manager is called to create a water splash rumble effect when a player lands in water from a height. Flying up to the sky, then starting to fall whilst holding the right stick to the right (causing the camera to constantly look to the right), then immediately letting go of the stick as the rumble begins signifying landing in water, the camera does not stop moving right. I have tracked down this behaviour right to the `SDL_GetGamepadAxis` call, where I print the state of the right stick x axis. It is continually printing, showing the driver does not stop polling SDL. This leads me to believe this is the fault of SDL.

## Further testing

I implemented a test rumble tone that goes like this: 
1. scale up the weak tone (over 1 second)
2. scale down the weak tone (over 1 second)
3. a 1/4 sec strong tone, then immediately a 1/4 sec weak tone, repeated 10 times
As previously observed, input stops working as I begin the test tone, however half way through (it's hard to pinpoint exactly where) input begins to work as normal again.

This issue was first observed on macOS 14.2.1 Sonoma, but a user has also experienced this same effect with Windows. Both were using a DualSense controller. As I am currently away, I have no other controllers to test this with. This could then be an issue with the DualSense hidraw driver, but it could also be a wider issue. I will update this thread if I get any test results from other controllers.

This also did not occur on SDL2.

## Context

This is the exact code I used to reproduce this
https://github.com/isXander/Controlify/blob/1.20.x/dev/src/main/java/dev/isxander/controlify/driver/sdl/SDL3GamepadDriver.java#L177
It's in Java, but the binding is 1:1 with the C library, and it wouldn't make any sense that the binding layer has anything to do with this.

**Comment by isXander** [comment]
Here's a response I just got back from my user:
> looks like it might be an issue specific to sony controllers, just tried a dualshock 4 which has the same problem but also tried a generic xbox controller which did not have the issue. Using ds4windows with the dualsense to emulate an xbox 360 controller works fine

**Comment by meyraud705** [comment]
Why are you using 0 for duration? Can you try with non zero value (like 100ms) and check if it still happen?

**Comment by isXander** [comment]
> Why are you using 0 for duration? Can you try with non zero value (like 100ms) and check if it still happen?

I use 0 as it seems to make it an infinite duration, since if the game is lagging and the tick rate is reduced, the rumble wouldn't flicker on and off. Setting it to 0 had caused no issues on SDL2 and I was under the impression it was intended behaviour.

**Comment by meyraud705** [comment]
When using 0, SDL will never tell the controller to stop rumble, but the controller will stop rumble by itself after some time (5s for Dualshock). You should use `SDL_MAX_RUMBLE_DURATION_MS` to get "infinite" rumble. Anyway it should not cause any issue.

I cannot reproduce this bug in `testgamecontroller` with Dualshock4 on Linux. Moving trigger to update rumble, all inputs update correctly. I will leave this to someone else, unless there is more info on how to trigger this.

**Comment by isXander** [comment]
Even with a 5 second duration it still happens.

**Comment by TheCrafter8000** [comment]
any update on this?

**Comment by isXander** [comment]
Not from me. Just waiting on SDL team.

**Comment by isXander** [comment]
I did some further testing:

I can now also repeat this on macOS Sonoma with a Stadia Controller connected via USB. I believe this can now rule out the DualSense controller firmware as a cause of this issue.

I have created some test vibration patterns with a duration of 2 seconds (40 ticks)

- 100% strong, 0% weak   - input freezes for the duration of 2 seconds
- 0% strong, 100% weak   - input freezes for the duration of 2 seconds
- 100% strong, 100% weak - input freezes for the duration of 2 seconds
- 50% strong, 0% weak    - input *does not* freeze whatsoever
- 50% strong, 50% weak   - input *does not* freeze whatsoever

I do have a theory to what is causing this:

I recently implemented the DS5 effect state to toggle on/off mute light. My game loop runs at 20 ticks per second; when calling `SDL_SendGamepadEffect` once every tick, the controller appears laggy, frozen. When only calling it when the mute light changes, it's fine.

My rumble manager supports dynamic effects which polls a new state from an effect instance every tick, and calls `SDL_RumbleGamepad` accordingly, even if the strong and weak values are identical (this is so the duration does not run out). My theory is that calling rumble so frequently is somehow lagging SDL. What I do not understand however is why a weaker effect does not trigger this, since the same amount of data is being sent.

**Comment by isXander** [comment]
I have tried to create a minimal example in C.

Copying the java code 1:1 into C, using
```c
while (1) {
	loop();
}
``` 
resulted in it working just fine.

However, when adding a sleep call here:
```c
while (1) {
	loop();
	sleep_ms(50); /* emulates 20tps game loop */
}
``` 
...all gamepad inputs return `0`, revealing the issue.

I wasn't able to test if the rumble causes an issue like this, since I could never get it to work in this state to test if rumble makes it stop working.

This issue slightly differs to what I am experiencing in my game, since with this minimal example inputs fail altogether, compared to only with certain rumble values and sending gamepad effects.

Here's the full code
```c
#include <SDL3/SDL.h>
#include <SDL3/SDL_main.h>
#include <SDL3/SDL_test.h>

#include "gamepadutils.h"
#include "testutils.h"

#ifdef WIN32
#include <windows.h>
#elif _POSIX_C_SOURCE >= 199309L
#include <time.h>   // for nanosleep
#else
#include <unistd.h> // for usleep
#endif

typedef struct
{
    SDL_JoystickID id;
    SDL_Gamepad *gamepad;

    Uint16 rumble_timer;
} Controller;

static Controller *controller;

void update_gamepad(Controller *controller)
{
    if (SDL_GetGamepadButton(controller->gamepad, SDL_GAMEPAD_BUTTON_SOUTH) == SDL_PRESSED) {
        controller->rumble_timer = 20;
    }

    Sint16 axis = SDL_GetGamepadAxis(controller->gamepad, SDL_GAMEPAD_AXIS_LEFTX);
    SDL_Log("Axis is %d", axis);

    /* this ends up being unused because SDL_GetGamepadButton never returns pressed */
    if (controller->rumble_timer > 0) {
        controller->rumble_timer--;
        SDL_RumbleGamepad(controller->gamepad, 0xFFFF, 0, 5000);
    }
}

void loop(void)
{
    if (controller != NULL) {
        update_gamepad(controller);
    }

    SDL_PumpEvents();

    SDL_Event event;
    while (SDL_PollEvent(&event)) {
        switch (event.type) {
        case SDL_EVENT_GAMEPAD_ADDED:
            Controller new_controller;

            SDL_zero(new_controller);

            new_controller.id = event.gdevice.which;
            new_controller.gamepad = SDL_OpenGamepad(new_controller.id);
            new_controller.rumble_timer = 0;

            controller = &new_controller;

            break;
        case SDL_EVENT_GAMEPAD_REMOVED:
            if (controller != NULL && controller->id == event.gdevice.which) {
                controller = NULL;
            }

            break;
        }
    }

    SDL_UpdateGamepads();
    SDL_UpdateJoysticks();
}

void sleep_ms(int milliseconds){
#ifdef WIN32
    Sleep(milliseconds);
#elif _POSIX_C_SOURCE >= 199309L
    struct timespec ts;
    ts.tv_sec = milliseconds / 1000;
    ts.tv_nsec = (milliseconds % 1000) * 1000000;
    nanosleep(&ts, NULL);
#else
    if (milliseconds >= 1000)
        sleep(milliseconds / 1000);
    usleep((milliseconds % 1000) * 1000);
#endif
}

int main(int argc, char *argv[])
{
    SDLTest_CommonState *state;

    /* Initialize test framework */
    state = SDLTest_CommonCreateState(argv, 0);
    if (!state) {
        return 1;
    }

    SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_PS3, "1");
    SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_PS4_RUMBLE, "1");
    SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_STEAMDECK, "1");

    /* Enable standard application logging */
    SDL_LogSetPriority(SDL_LOG_CATEGORY_APPLICATION, SDL_LOG_PRIORITY_INFO);

    if (SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_GAMEPAD | SDL_INIT_EVENTS | SDL_INIT_AUDIO) < 0) {
        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't initialize SDL: %s\n", SDL_GetError());
        return 1;
    }

    SDL_AddGamepadMappingsFromFile("gamecontrollerdb.txt");

    int gcount;
    SDL_JoystickID *gamepads = SDL_GetGamepads(&gcount);
    if (gcount > 0) {
        SDL_Log("Found initial controller");

        SDL_JoystickID jid = gamepads[0];

        Controller new_controller;

        SDL_zero(new_controller);

        new_controller.id = jid;
        new_controller.gamepad = SDL_OpenGamepad(jid);

        controller = &new_controller;
    }

    while (1) {
        loop();
        sleep_ms(50); /* emulates 20tps game loop */
    }
}

```
This is my first time writing C code, so apologies for the bad looking code.

**Comment by slouken** [comment]
@isXander, it looked like you created a test program to reproduce the issue and then deleted it? A test program would be really helpful here to reproduce the problem.

**Comment by isXander** [comment]
> @isXander, it looked like you created a test program to reproduce the issue and then deleted it? A test program would be really helpful here to reproduce the problem.

The test program I made had a bug in it that displayed a similar issue. Once I fixed it, I couldn't repeat it. So I deleted it since I couldn't replicate the issue.

I'm currently exploring potential solutions that might be (yet another) JNA issue. Will keep this thread updated with any more stuff that I find.

**Comment by svenvvv** [comment]
I encountered a very similar problem with SDL3 (version `preview-3.1.6`, built from source tarball) under Linux. When rumble is enabled in the game then it seems like there is a very large delay in processing input events.

I'm also able to reproduce a similar problem with the `testcontroller` program: the input values are updated correctly there, but the rumble effects are greatly delayed. The rumble effects seem to apply approximately every 5 seconds, instead of reacting immediately to the trigger states.
E.g. tapping the trigger once causes a rumble that lasts 5 seconds, pressing the trigger after it turned off doesn't take effect until another 5 seconds have passed from the previous state change.

The rumble problems are present with an original DualShock 3 (PS3) controller, both over Bluetooth and when wired.
Also tested with a Logitech RumblePad 2 USB and it has none of the problems I mentioned and works as expected.

**Comment by svenvvv** [comment]
I did a bit of debugging and I observed the `hid_write` (which basically just calls Linux `write()`) taking five seconds every time when called from `SDL_HIDAPI_RumbleThread`. Went back to try with SDL2 and it didn't work well there either.
I'm not really sure why or what exactly broke, but restarting `bluetooth.service` fixed the problem, now the DS3 controller works fine with `testcontroller`.
When I run the game I was updating to SDL3 then it breaks again, but I'm not sure that it's caused by SDL or just a coincidence.

**Comment by TinLikesCode** [comment]
has this been fixed? I cant use my xbox controller for some reason because it wont work properly, only the triggers are working so I have to resort to my duelsense controller, this bug has been a primary issue for me

**Comment by slouken** [comment]
> has this been fixed? I cant use my xbox controller for some reason because it wont work properly, only the triggers are working so I have to resort to my duelsense controller, this bug has been a primary issue for me

Does it work normally if you run "Set up USB game controllers" and then click Properties to test the Xbox controller?

**Comment by isXander** [comment]
> has this been fixed? I cant use my xbox controller for some reason because it wont work properly, only the triggers are working so I have to resort to my duelsense controller, this bug has been a primary issue for me

You are describing another issue present with my mod, I'd recommend sticking to the github issues for my mod, not the underlying native library with issues that are vaguely similar.

**Comment by isXander** [comment]
> I did a bit of debugging and I observed the `hid_write` (in `src/hidapi/linux/hid.c`, which basically just calls Linux `write()`) taking five seconds every time when called from `SDL_HIDAPI_RumbleThread`. Went back to try with SDL2 and it didn't work well there either. I'm not really sure why or what exactly broke, but restarting `bluetooth.service` fixed the problem, now the DS3 controller works fine with `testcontroller`. When I run the game I was updating to SDL3 then it breaks again, but I'm not sure that it's caused by SDL or just a coincidence.

This seems close to my issue, but what I find extremely weird is that it's only on more intense vibrations, using a magnitude of 0x50 completes the vibration successfully, without any issue, but using a magnitude of 0xFF, makes input freeze completely until the vibration stops.

**Comment by slouken** [comment]
> > I did a bit of debugging and I observed the `hid_write` (in `src/hidapi/linux/hid.c`, which basically just calls Linux `write()`) taking five seconds every time when called from `SDL_HIDAPI_RumbleThread`. Went back to try with SDL2 and it didn't work well there either. I'm not really sure why or what exactly broke, but restarting `bluetooth.service` fixed the problem, now the DS3 controller works fine with `testcontroller`. When I run the game I was updating to SDL3 then it breaks again, but I'm not sure that it's caused by SDL or just a coincidence.
> 
> This seems close to my issue, but what I find extremely weird is that it's only on more intense vibrations, using a magnitude of 0x50 completes the vibration successfully, without any issue, but using a magnitude of 0xFF, makes input freeze completely until the vibration stops.

Are you able to reproduce this with testcontroller? Pulling the trigger all the way back should generate a magnitude 0xFFFF rumble (which is what I'm guessing you're intending here?)

**Comment by isXander** [comment]
> Are you able to reproduce this with testcontroller? Pulling the trigger all the way back should generate a magnitude 0xFFFF rumble (which is what I'm guessing you're intending here?)

I have never been able to reproduce this consistently, nor ever with testcontroller. 

To cast the net a bit wider, here are some configurations where this has happened:
- All major operating systems - Windows, macOS, Linux
- Many controllers - Dualshock4 (PS4), Dualshock3 (PS3), Dualsense (PS5), Stadia, Switch Pro, Aftermarket Dualshock3
- Both USB and Bluetooth

**Comment by slouken** [comment]
Maybe CPU starvation of the HIDAPI rumble thread?

**Comment by isXander** [comment]
> Maybe CPU starvation of the HIDAPI rumble thread?

Will this semi-confirm this if I confirm this doesn't happen with controllers that use x-input, rawinput, or dinput, like xbox-adjacent controllers?

**Comment by slouken** [comment]
> > Maybe CPU starvation of the HIDAPI rumble thread?
> 
> Will this semi-confirm this if I confirm this doesn't happen with controllers that use x-input, rawinput, or dinput, like xbox-adjacent controllers?

Not really. You could try increasing the thread priority in the rumble thread and see if that makes it happen less frequently. I can post a patch here if you need it.

**Comment by isXander** [comment]
Just a quick update: this is definitely still happening in the latest commits of SDL3. I am also yet to see a single person who doesn't use HIDAPI drivers experience this.

It's a really troublesome issue, as I myself cannot get it to repeat, even inside of my game where these reports are coming from.

**Comment by slouken** [comment]
As an experiment you can remove the rumble thread and make the HID write call inline. This is definitely something that we can't ship because the thread was added because there are lots of cases where the write will block for a long time, but at least we can tell whether it's a scheduling issue.

**Comment by isXander** [comment]
> As an experiment you can remove the rumble thread and make the HID write call inline. This is definitely something that we can't ship because the thread was added because there are lots of cases where the write will block for a long time, but at least we can tell whether it's a scheduling issue.

Yes, I'll try and gather a test group of people who experience this. It might take you a few days to get some results.

**Comment by isXander** [comment]
@slouken Indeed, when doing it in the main thread, the locking up goes away. A bit preliminary since I have only had a single user test it so far, but seems this is it. Check the issue I referenced yesterday.

**Comment by isXander** [comment]
Now that the cause has been discovered, will this be added back to the 3.2 milestone?

**Comment by isXander** [comment]
## How the issue manifests

Note: this explanation is preliminary, and has not actually been confirmed, but it seems pretty clear that this is what it must be.

From what I understand, inside `SDL_HIDAPI_SendRumble*()`, the driver bumps an atomic `rumble_pending` counter and wakes the dedicated SDL rumble thread. The thread then
```c
SDL_LockMutex(joystick_lock);
hid_write(device, out_report, len);   /* can block 5-20 ms over BT */
--rumble_pending;
SDL_UnlockMutex(joystick_lock);
SDL_Delay(10);
```
(simplified code)

Because it keeps the same mutex that `SDL_GetGamepadAxis()` needs (and other input methods), every write stalls input for the entire duration of the HID write the thread-slice it monopolises. Under a flood of rumble updates that my mod submits, the delay snowballs.

`rumble_pending` is a simple incrementing counter, if it's enqueued faster than the thread can dequeue (worse over BT), plus SDL's own PS5 "tickle" packets, the backlog grows without bound. This eventually leads to waiting of thousands of writes.

The reason the high priority thread priority doesn't work is because when the rumble queue is huge, the read thread is ready, but starved for mutex time. Changing priorities won't help, since they fight over the same lock.

## Proposed solutions

I think that we could implement solution 3 as a stop-gap during `3.2.x`, then a more potentially-breaking patch like solution 1 could be implemented in `3.3.0`?

### 1. Replace `rumble_pending`

Rumble is not something that needs a huge queue of updates that are dispatched all at once. When the mutex releases and we can write the rumble HID, use the *latest* rumble state, don't go through the whole queue; controllers can't react that fast anyway. All we need is the latest strong/weak pair.

Replace the counter with a ring buffer of size 1: when the main thread queues a new rumble state, it should *overwite* the previous one, not incrementing a counter. The RumbleThread wakes, sends one packet, goes back to sleep, no uncontrolled backlog.

### 2. Separate locks

Split the joystick lock, give each HIDAPI device its own mutex for the long blocking writes. Then input and rumble stop contending.

### 3. Refuse additional rumble events

Return `SDL_EAGAIN` if the caller tries to queue a second rumble packet when one is still pending (`rumble_pending > 0`).

**Comment by isXander** [comment]
Hey, could I get some feedback on my proposed solutions? I'd be happy to attempt at making a PR for whatever solution you guys deem best.

**Comment by meyraud705** [comment]
> `rumble_pending` is a simple incrementing counter, if it's enqueued faster than the thread can dequeue (worse over BT), plus SDL's own PS5 "tickle" packets, the backlog grows without bound. This eventually leads to waiting of thousands of writes.

The queue cannot grow that much:
 - Tickle packet is only sent if we do not read any packet for `BLUETOOTH_DISCONNECT_TIMEOUT_MS` (500ms). 
 - The rumble thread updates the pending request with value from new request instead of enqueuing it, if possible. https://github.com/libsdl-org/SDL/blob/23e08f780785fccf30fdd19b775e43b26434f313/src/joystick/hidapi/SDL_hidapi_rumble.c#L265-L271

**Comment by slouken** [comment]
@meyraud705 is correct, the queue shouldn't be getting that long. Can you confirm that you're actually seeing large queue lengths?

The rumble and report reading need to take the same lock, as on most operating systems you can't read and write the same file descriptor from different threads at the same time.

**Comment by AL2009man** [comment]
I asked at least one person who'd done Input stuffs on Sourceports about this, but their contributions will be limited. They have told me similar issues has occurred with Nintendo Switch controllers. Recently: they did tested on both yquake2 and woof sourceports to confirm this bug isn't affected...

but long-story short: this problem becomes far more evident when using a Nintendo Switch Joy-Cons on a bad Bluetooth Adapter or Motherboard. a decent or good Bluetooth Adapter won't exhibit such issues.

**Comment by isXander** [comment]
> but long-story short: this problem becomes far more evident when using a Nintendo Switch Joy-Cons on a bad Bluetooth Adapter or Motherboard. a decent or good Bluetooth Adapter won't exhibit such issues.

My users and I have observed this even on wired USB connections.

**Comment by isXander** [comment]
I'd love to get this fixed.
I've managed to observe this bug with a debugger attached to SDL.
Any pointers on where I might start?

**Comment by isXander** [comment]
## Findings

This "freeze" is immediately resolved by sending an additional `SDL_RumbleJoystick(ptr, 0,0)`. So whatever cleanup it does, gets it out the bad state.


As we know rumble thread and joystick thread share a `dev_lock` mutex. This is the cause.
https://github.com/libsdl-org/SDL/blob/ca537d651b82f28bc280685391665eb25c856e00/src/joystick/hidapi/SDL_hidapijoystick.c#L1428-L1444
During the observed "freeze", the main HIDAPIjoystick update method cannot acquire the mutex and therefore updates are skipped.
This was observed by attaching a logging breakpoint within, and I observed that during the freeze the printing stopped.
  

HID writes on the rumble thread can take up to 3ms, but this proves that it isn't hanging on a single hid_write, but a scheduler starvation issue.
```
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=3.58ms
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=1.62ms
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=3.00ms
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=0.58ms
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=2.55ms
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=2.34ms
Rumble write: device=000002084aaedec0 size=48 result=48 lock_wait=0.00ms write_time=0.24ms
```


The rumble thread has a high thread priority. It seems that scheduler starvation is a very likely cause. This explains why it is so inconsistent in producibility, because it's up to the OS scheduler on how it handles threads.


In my usage of SDL where I am observing this bug, rumble requests are sent at a relatively low rate of 20Hz, which I would have thought would not have been enough to starve the other thread.
Especially since the rumble thread also waits 10ms between each update as a measure to attempt to not starve other threads.

This might suggest that it is not the rumble thread that is holding the lock.

It is not the main HIDAPI update joystick that is holding the lock. No locks ever take longer than 200ms (that was the threshold I has a log at).

Clear starvation evidence below:
```
HIDAPI_UpdateDevices: calls=84/s lock_success=84/s lock_fail=0/s slow(>2ms)=0/s max_update=0.417ms
Rumble requests in last second: 0
HIDAPI_UpdateDevices: calls=84/s lock_success=84/s lock_fail=0/s slow(>2ms)=0/s max_update=0.497ms
Rumble requests in last second: 0
HIDAPI_UpdateDevices: calls=84/s lock_success=84/s lock_fail=0/s slow(>2ms)=0/s max_update=0.407ms
Rumble requests in last second: 0
HIDAPI_UpdateDevices: calls=80/s lock_success=80/s lock_fail=0/s slow(>2ms)=0/s max_update=0.638ms
Rumble requests in last second: 1
HIDAPI_UpdateDevices: calls=84/s lock_success=31/s lock_fail=53/s slow(>2ms)=0/s max_update=0.227ms
Rumble requests in last second: 21
HIDAPI_UpdateDevices: calls=80/s lock_success=0/s lock_fail=80/s slow(>2ms)=0/s max_update=0.000ms
Rumble requests in last second: 20
HIDAPI_UpdateDevices: calls=80/s lock_success=0/s lock_fail=80/s slow(>2ms)=0/s max_update=0.000ms
Rumble requests in last second: 21
HIDAPI_UpdateDevices: calls=80/s lock_success=4/s lock_fail=76/s slow(>2ms)=0/s max_update=0.481ms
Rumble requests in last second: 20
HIDAPI_UpdateDevices: calls=84/s lock_success=0/s lock_fail=84/s slow(>2ms)=0/s max_update=0.000ms
Rumble requests in last second: 20
HIDAPI_UpdateDevices: calls=84/s lock_success=0/s lock_fail=84/s slow(>2ms)=0/s max_update=0.000ms
Rumble requests in last second: 21
HIDAPI_UpdateDevices: calls=84/s lock_success=0/s lock_fail=84/s slow(>2ms)=0/s max_update=0.000ms
Rumble requests in last second: 20
HIDAPI_UpdateDevices: calls=80/s lock_success=41/s lock_fail=39/s slow(>2ms)=0/s max_update=0.483ms
Rumble requests in last second: 0
HIDAPI_UpdateDevices: calls=84/s lock_success=84/s lock_fail=0/s slow(>2ms)=0/s max_update=0.187ms
Rumble requests in last second: 0
HIDAPI_UpdateDevices: calls=80/s lock_success=80/s lock_fail=0/s slow(>2ms)=0/s max_update=0.177ms
Rumble requests in last second: 0
HIDAPI_UpdateDevices: calls=84/s lock_success=84/s lock_fail=0/s slow(>2ms)=0/s max_update=0.198ms
```

I tracked the owner of the lock, and printed the owner when HIDAPI_UpdateDevices fails. It's always the rumble write that owns the dev_lock mutex. This makes sense. I also timed how long it takes the rumble thread to acquire the mutex, it's always 0.00ms. It's always available when rumble asks for it.


Despite reducing the thread priority of the rumble thread to normal, starvation does not lessen and all 80 updates per second fail.


SDL calls HIDAPI_UpdateDevices at 80Hz, which is every ~12.5ms. The rumble thread waits for a semaphore signal, if there is a request, it holds the mutex, hid_write, unlocks the mutex, and waits 10ms.
Maybe there's some phase matching that means rumble always gets there first?


Over each second, the rumble thread is only holding the lock for ~40ms in total. Yet HIDAPI_UpdateDevices TryLock fails ~80/80 times?
```
Rumble req/s=20 lock_held_time=43.67ms avg_lock_held_time=2.18ms
```

Calls to `HIDAPI_UpdateDevices` are bursty. Yes, the rumble thread only locks for ~40ms/second. But in my case, `HIDAPI_UpdateDevices` calls do not happen every ~12.5ms like 80req/sec suggests, but all of the failed calls happen within <0.001ms of eachother.

The bursty calls were the fault of my game that uses SDL. After fixing this (I was causing UpdateJoysticks to happen more than once per tick), the bug is still prevelant, and now we're at HIDAPI_UpdateDevices to the exact same as rumble: 20req/s.
`UpdateDevices timing: min=42.499ms avg=49.883ms max=52.289ms`

**Comment by isXander** [comment]
## Identified cause / effect

So this was indeed an issue of phase. 
After I removed the duplicate calls to UpdateJoystick, it became clear that joysticks were being updated at the same interval as new rumble states were being set. 20Hz.

1. `SDL_RumbleJoystick` -> triggers semaphore -> locks dev_lock mutex
2. <4ms goes by (the avg. time rumble thread locks mutex for)
3. `SDL_PollEvent` -> `SDL_PumpEvents` -> `SDL_UpdateJoysticks` -> `HIDAPI_UpdateDevices`
4. `HIDAPI_UpdateDevices` fails, because we're in the middle of sending rumble effect
5. game loop continues, game thinks no events have been polled, updates rumble effect again. loop repeats

## Quick fix

Adding 0-2ms jitter to the rumble thread after the semaphore is signalled, completely dissolves the issue. The failed mutex locks goes down to 0.

Obviously, adding a random jitter is not PR-quality. I am now assessing proper solutions.

**Comment by isXander** [comment]
## Deterministic starvation under periodic contention

There are no measures in place to prevent this.

`HIDAPI_UpdateDevices()` uses` SDL_TryLockMutex(device->dev_lock)` and skips the update entirely if the lock is held.

The HID output (rumble and other device writes) is performed on a dedicated "rumble" thread which briefly locks the same mutex for each write.

Individually, these lock holds are short (~1–3 ms), and the lock is free most of the time. However, if an application drives both:
- event pumping (`SDL_PumpEvents` -> `HIDAPI_UpdateDevices`)
- rumble/output submission (`SDL_RumbleGamepad` -> rumble thread)

from a fixed-rate game loop, the two operations become periodic and phase-locked.

Because `HIDAPI_UpdateDevices` attempts the lock only once per tick, if its TryLock happens to consistently fall inside the rumble thread’s brief lock window, input updates will be skipped every tick -- even though the lock is free for the majority of the frame. This creates a deterministic livelock/starvation condition where joystick state never updates while rumble is active.

The problem is not that the mutex is always locked - it is that the reader samples at a single deterministic instant and never retries, so a stable phase alignment can produce indefinite starvation.

Introducing any phase-breaking mechanism (random jitter, yield-and-retry, or short bounded waiting) immediately resolves the issue, demonstrating that the root cause is lack of fairness in the TryLock-and-skip strategy.

## Unbounded request

The request queue is unbounded, yet we wait 10ms after processing a single request.

This is potentially problematic, as if badly behaving users of SDL spam effect updates, we get and effective latency of ~2ms + 10ms (time for write to complete and anti-starvation delay) multiplied by the request queue length. That's a drain rate of just 84 req/sec.

If the user's game loop has a high Hz, this can quite quickly lead to a runaway queue length.

**Comment by slouken** [comment]
Nice work identifying the issue! Feel free to submit a PR to fix this.

**Comment by isXander** [comment]
> Nice work identifying the issue! Feel free to submit a PR to fix this.

Problem is, it's not trivial to fix. How do we ensure that reading gets time with the mutex?

**Comment by isXander** [comment]
Further evidence that I am correct in my evaluation.

This is a trace file I generated.
Open in https://ui.perfetto.dev/
You will see two tracks, a point where HIDAPI_UpdateDevices attempts to take the mutex, and a duration, where rumble thread is holding the mutex.

[hid_trace.json](https://github.com/user-attachments/files/24555084/hid_trace.json)

<img width="2317" height="274" alt="Image" src="https://github.com/user-attachments/assets/2e63deb8-91e7-4f7d-bac6-950c7c1eff24" />

**Comment by isXander** [comment]
If you could let me know *how* you'd like to address this, I'd be happy to PR.

**Comment by slouken** [comment]
Yield and retry for a short bounded time is probably the best approach here. We don't want to get into a case where we're hung for a long time on I/O, but we don't want to lose input either.

**Comment by isXander** [comment]
> Yield and retry for a short bounded time is probably the best approach here. We don't want to get into a case where we're hung for a long time on I/O, but we don't want to lose input either.

If you look at the trace graph, you'll see that TryLock happens (in my case anyway) right at the beginning of rumble holding the mutex. So we'd have to delay by >2ms on the main thread in order to actually succeed. And this way also doesn't enforce that it will ever succeed. I believe some explicit logic should be in place to ensure we don't starve input, not by just playing the numbers game.

**Comment by isXander** [comment]
I'm pretty sure that you can read and write the same HID device at the same time, since I believe read is purely a dequeue anyway, at least on hidraw?
Maybe the `dev_lock` could be split into a `write_lock` and `read_lock`?

Other ideas I've thought of, an epoch on the device that counts the amount of successful device updates, that way we can deterministically load balance reads and writes.

**Comment by isXander** [comment]
An easy reproduction of this is simply this:
```c
while (true) {
	SDL_RumbleGamepad(ptrGamepad, random(), random()); // using random for magnitude so SDL doesn't cache and prevent sending a write
	SDL_UpdateGamepads();

	SDL_Delay(50);
}
```
The read within `SDL_UpdateGemapds()` will fail, because the rumble thread will have the mutex.

**Comment by slouken** [comment]
> I'm pretty sure that you can read and write the same HID device at the same time, since I believe read is purely a dequeue anyway, at least on hidraw? Maybe the `dev_lock` could be split into a `write_lock` and `read_lock`?

It depends on the platform and driver. On Windows some devices will actually crash if you try to read and write simultaneously.

**Comment by slouken** [comment]
> An easy reproduction of this is simply this:
> 
> while (true) {
> 	SDL_RumbleGamepad(ptrGamepad, random(), random()); // using random for magnitude so SDL doesn't cache and prevent sending a write
> 	SDL_UpdateGamepads();
> 
> 	SDL_Delay(50);
> }
> The read within `SDL_UpdateGemapds()` will fail, because the rumble thread will have the mutex.

What platform and controller are you testing with this repro case?

**Comment by isXander** [comment]
> What platform and controller are you testing with this repro case?

Any HIDAPI controller driver, I've had this occur on all major OSes.

The repro is purely theoretical, I haven't actually ran that specific code. You may need a small delay between rumble and UpdateGamepads to give the rumble thread a chance to wake up. That delay simulates some code running in the game loop between the two calls.

**Comment by meyraud705** [comment]
What about signalling the rumble thread right after a read in `HIDAPI_UpdateDevices()`? It adds a delay to rumble but there is no missing read unless the write takes too long.

```diff
diff --git a/src/joystick/hidapi/SDL_hidapi_rumble.c b/src/joystick/hidapi/SDL_hidapi_rumble.c
index ef3fede52..039dfdd4f 100644
--- a/src/joystick/hidapi/SDL_hidapi_rumble.c
+++ b/src/joystick/hidapi/SDL_hidapi_rumble.c
@@ -234,11 +234,8 @@ int SDL_HIDAPI_SendRumbleWithCallbackAndUnlock(SDL_HIDAPI_Device *device, const
     }
     ctx->requests_head = request;
 
-    // Make sure we unlock before posting the semaphore so the rumble thread can run immediately
     SDL_HIDAPI_UnlockRumble();
 
-    SDL_SignalSemaphore(ctx->request_sem);
-
     return size;
 }
 
@@ -273,6 +270,19 @@ int SDL_HIDAPI_SendRumble(SDL_HIDAPI_Device *device, const Uint8 *data, int size
     return SDL_HIDAPI_SendRumbleAndUnlock(device, data, size);
 }
 
+void SDL_HIDAPI_UpdateRumble(void)
+{
+    SDL_HIDAPI_RumbleContext *ctx = &rumble_context;
+    if (SDL_GetAtomicInt(&ctx->running)) {
+        SDL_LockMutex(SDL_HIDAPI_rumble_lock);
+        SDL_HIDAPI_RumbleRequest *request = ctx->requests_tail;
+        SDL_UnlockMutex(SDL_HIDAPI_rumble_lock);
+        if (request) {
+            SDL_SignalSemaphore(ctx->request_sem);
+        }
+    }
+}
+
 void SDL_HIDAPI_QuitRumble(void)
 {
     SDL_HIDAPI_RumbleContext *ctx = &rumble_context;
diff --git a/src/joystick/hidapi/SDL_hidapi_rumble.h b/src/joystick/hidapi/SDL_hidapi_rumble.h
index bcd4a790d..86f7ea312 100644
--- a/src/joystick/hidapi/SDL_hidapi_rumble.h
+++ b/src/joystick/hidapi/SDL_hidapi_rumble.h
@@ -39,4 +39,7 @@ void SDL_HIDAPI_UnlockRumble(void) SDL_RELEASE(SDL_HIDAPI_rumble_lock);
 int SDL_HIDAPI_SendRumble(SDL_HIDAPI_Device *device, const Uint8 *data, int size);
 void SDL_HIDAPI_QuitRumble(void);
 
+// Tell rumble thread to write pending rumbles to devices
+void SDL_HIDAPI_UpdateRumble(void);
+
 #endif // SDL_JOYSTICK_HIDAPI
diff --git a/src/joystick/hidapi/SDL_hidapijoystick.c b/src/joystick/hidapi/SDL_hidapijoystick.c
index 4ae31e270..76d3daad6 100644
--- a/src/joystick/hidapi/SDL_hidapijoystick.c
+++ b/src/joystick/hidapi/SDL_hidapijoystick.c
@@ -1440,6 +1440,7 @@ void HIDAPI_UpdateDevices(void)
             }
         }
         HIDAPI_FinishUpdatingDevices();
+        SDL_HIDAPI_UpdateRumble();
     }
 }
 
@@ -1670,7 +1671,7 @@ static bool HIDAPI_JoystickSetSensorsEnabled(SDL_Joystick *joystick, bool enable
 
 static void HIDAPI_JoystickUpdate(SDL_Joystick *joystick)
 {
-    // This is handled in SDL_HIDAPI_UpdateDevices()
+    // This is handled in HIDAPI_UpdateDevices()
 }
 
 static void HIDAPI_JoystickClose(SDL_Joystick *joystick) SDL_NO_THREAD_SAFETY_ANALYSIS // We unlock the device lock so rumble can complete
@@ -1686,6 +1687,7 @@ static void HIDAPI_JoystickClose(SDL_Joystick *joystick) SDL_NO_THREAD_SAFETY_AN
             // Unlock the device so rumble can complete
             SDL_UnlockMutex(device->dev_lock);
         }
+        SDL_HIDAPI_UpdateRumble();
         for (i = 0; i < 3; ++i) {
             if (SDL_GetAtomicInt(&device->rumble_pending) > 0) {
                 SDL_Delay(10);
```
